cszczepaniak/cribbly-frontend
# Reroll auth [frontend]

# What I added (link any issues addressed)
![cribbly-auth](https://user-images.githubusercontent.com/19275178/104084342-b0d81080-520b-11eb-8306-752e8516f9b4.gif)

Set up auth with Firebase auth instead of Auth0. After a bit of struggle, I got things working and I have to say, it's pretty seamless. I took a ton of inspiration (read: I mostly copied) a custom react hook I found [here](https://usehooks.com/useAuth/). I modified things for our use case. I love custom hooks!!

After that, it was just about setting up the pages to call into the custom hooks. Another big change with this PR is seen on the welcome page; there's a `useEffect` that calls into the Login endpoint I set up a while back on the backend to create or get the user (depending on if the user existed or not).

The cool thing is, it will be absolutely trivial for us to add email and password login, Facebook login, or any other social provider. My preference would be for using _only_ social logins in prod (because then users won't have to manually type in their name, and we don't have to worry as much about HTTPS), but in dev maybe we should add email and password so it's easy to create test users.

# How to test it
- Make sure you have the frontend and backend checked out to their respective `rerollAuth` branches
- You'll need to grab the firebase config and put it in `src/secrets/firebaseConfig.json` as shown below. You can grab the config with values filled in from firebase if you go to the project settings, then scroll down to the cribbly-dev web app, then grab the firebase SDK snippet.
![image](https://user-images.githubusercontent.com/19275178/104096230-dc391a80-5260-11eb-96a6-f351c542e825.png)

```json
{
    "apiKey": "the-api-key",
    "authDomain": "the-domain",
    "projectId": "the-project-id",
    "storageBucket": "the-storage-bucket",
    "messagingSenderId": "the-messaging-sender-id",
    "appId": "the-app-id"
}
```
- You'll also need to add a dotnet user secret that has the firebase audience set. I'm not sure if this is secret or not, but I decided to be safe. Use `dotnet user-secrets set FirebaseAuth:Audience "the-audience" --project CribblyBackend`. The audience is the project ID also found in the firebase project settings.
- Make sure MySQL is running
- Spin up the backend
- Spin up the frontend
- Try clicking the "Access Private Data" button on the frontend; you should get a 401
- Log in (should be with google)
- You should be greeted by name; it should say "welcome" the first time you login and "welcome back" subsequent times
- Try clicking the "Access Private Data" button on the welcome page; you should get "Hello, private cribbly!"
